rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,496 @@
+VNet and TCP app access in such a way that the end user can connect to
+`tcp-app.teleport.example.com:1234` and the connection will be forwarded to the same port number on
+the hostname specified in the app spec.
+
+## Details
+
+### UX
+
+#### Setup
+
+There's an app called waldo which accepts connections on multiple ports. Alice wants to make the app
+available in the teleport.example.com cluster. She creates a new app and includes a new field called
+`ports` in the app definition. She leaves the port number out of the `uri` field.
+
+```yaml
+apps:
+- name: "waldo"
+  uri: "tcp://localhost"
+  ports: [4080-4090, 5095]
+  labels:
+    env: prod
+```
+
+#### Usage
+
+##### Happy path
+
+###### Regular usage
+
+Bob wants to connect from his device to waldo using a CLI client called waldo-client. Bob sees in
+Teleport Connect that the app waldo.teleport.example.com supports ports 4080 and 4081, which are the
+default ports over which waldo-client connects. Bob starts VNet, opens his terminal of choice and
+points waldo-client at the app in the cluster.
+
+```
+waldo-client waldo.teleport.example.com
+```
+
+###### Local proxy
+
+Charlie needs to connect to waldo's debug service. Its API is typically available over the port
+5095. Charlie uses Linux, where VNet is not yet available. Through `tsh apps ls` they see that the
+waldo app supports the port 5095, so they start a local proxy that targets that port:
+
+```
+tsh proxy app waldo --target-port 5095 --port 5095
+waldo-debug localhost # In a separate shell session.
+```
+
+##### Error path
+
+###### Incorrect port
+
+Dave wants to connect to the debug port through VNet, but he makes a typo and provides a port
+number that's not in the app spec.
+
+```
+waldo-debug waldo.teleport.example.com:5096
+```
+
+The connection reaches the app service which closes the connection because of an unknown port.
+waldo-debug observes the connection being unexpectedly closed. Dave doesn't see any immediate
+information about an incorrect port – the app service has no way of passing this information as it
+does not know about the protocol used by waldo. Dave has to ask Alice to look into the logs of the
+app agent. She notices that around the time when the connection was made, the app service logged a
+warning saying that port 5096 is not defined in the app spec.
+
+```
+2024-08-23T16:47:02+02:00 WARN [APP:SERVI] Failed to handle client connection. error:[
+ERROR REPORT:
+Original Error: *trace.BadParameterError port 5096 not found in ports of the app "waldo"
+Stack Trace: …
+```
+
+###### Invalid app spec
+
+Alice wants to extend the app spec with another port, 6060, but she makes a mistake and forgets a
+comma:
+
+```yaml
+apps:
+- name: "waldo"
+  uri: "tcp://localhost"
+  ports: [4080-4090, 50956060]
+  labels:
+    env: prod
+```
+
+The relevant UI (`teleport start`, `tctl edit`, etc.) returns an error about an incorrect port
+in the app spec and prevents Alice from saving such app spec.
```

Comment:
We can reject the update if it's dynamic configuration (`tctl edit`), but what do we do if this is a teleport.yaml? We could:
- Prevent the entire app service from starting, but this could cause users to lose access to other apps
- Start the app service without the `waldo` app
- Start the app service with waldo but only for ports 4080-4090

Reply:
Currently, teleport fails to start with an error if you have an invalid app spec in `teleport.yaml` (e.g. you [put a port number in `public_addr`](https://github.com/gravitational/teleport/blob/ca136609184db1958e558a6b152db0f36e84cad9/lib/service/servicecfg/app.go#L116-L118)). I was thinking of preserving this behavior.